dotnet/src/SemanticKernel.Abstractions/AI/FunctionChoiceBehaviors/AutoFunctionChoiceBehavior.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Represent <see cref="FunctionChoiceBehavior"/> that provides either all of the <see cref="Kernel"/>'s plugins' function information to the model or a specified subset.
+/// This behavior allows the model to decide whether to call the functions and, if so, which ones to call.
+/// </summary>
+[Experimental("SKEXP0001")]
+public sealed class AutoFunctionChoiceBehavior : FunctionChoiceBehavior
```

Comment:
There are two reasons, both related to the deserialization aspect of the function choice behavior classes:

**1. Polymorphic deserialization:**
The function choice behavior classes are deserialized from JSON and YAML formats polymorphically. As such, deserialization API need to know about their types to be able to create their instances.  System.Text.Json.JsonSerializer can get the information from `JsonDerivedType` attribute the `FunctionChoiceBehavior` base class is annotated with. While the `YamlDotNet.Serialization.IDeserializer` implementation doesn't know anything about those attributes and requires a custom registration of the classes on the destabilizer - [WithTypeDiscriminatingNodeDeserializer](https://github.com/microsoft/semantic-kernel/blob/a3527d7aadb4ad75db437da8f6a3da3041a95ef4/dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs#L31). The YAML prompts deserialization logic resides in the `Functions.Yaml` project and needs the function choice behavior classes that live in the `SemanticKernel.Abstractions` project to be public to access them.

A solution here could be either:
**a.** Not to use polymorphic deserialization 
**b.** YAML deserialization functionality can read the function choice behavior types from the `JsonDerivedType` attribute the `FunctionChoiceBehavior` base class is annotated with using the `System.Reflection.GetCustomAttribute` method and cache it.

**2. Deserialization of function names:** SK prompt templates can call functions specified in the templates using syntax that uses '.' as a separator between plugin and function names. While configuring function choice behavior in the same template, it makes sense to use the same separator. In the example below, both function references use '.' as a separator:  
   
```yaml  
template: Say hello world in {{languageUtils.getLanguage}} to the current user  
execution_settings:  
  service1:  
    temperature: 1.0  
    function_choice_behavior:  
      type: auto  
      functions:  
        - graph.getCurrentUser  
```  
   
However, there could potentially be custom prompt templates that use '_', '+', '-', etc., as a separator in both the prompt template and function choice behavior. 

The solution here could be either:  
**a.** Normalize different separators to the well-known one '-' at the deserialization phase. This solution is used in this PR and requires access to the `Functions` collection of the function choice behavior classes that need to be public so YAML type converters can access them to do the separator normalization.  
**b.** Add a new `FunctionNameSeparator` property to the `FunctionChoiceBehavior` class so the separator can be specified in the prompt. This way, no normalization is required because the separator will be set explicitly if it differs from the default prompt one, '.'.

I can try to prototype the 1.b & 2.b solutions to make the function choice behavior classes internal.

Reply:
Fixed by implementing solution 1.b and using '.' as the default separator for plugin and function names. If, later on, there is a need to support a different separator for a new prompt template format, it can be achieved either by solution 2.b or by custom deserialization of the new prompt template.